### PR TITLE
Covariant return updates to ecma augments

### LINKED ISSUE #37265: Fixes to the covariant returns support
This issue tracks changes we will likely want to make to the covariant returns support after the #35308 merge.

- [x] Visit the pointers, byrefs, function pointer returns scenarios and determine if they should be supported, and add test coverage.
- [ ] Add a chapter on the covariant returns to https://github.com/dotnet/runtime/blob/master/docs/design/specs/Ecma-335-Augments.md ? Try to match the style and language used by ECMA-335 (e.g. in the ideal case - we would just copy&paste it over once we figure out how to make edits to ECMA-335).
- [x] Crossgen2 - for performance reasons, we will likely need to burn a MethodFlags bit for RequiresSlotUnification (in the AttributeMetadataCache category) and make this a proper member on MethodDesc.

### LINKED ISSUE #43814: ECMA-335 Augments documentation are missing some changes.
I asked in the C# discord recently about if the ECMA-335 specification was up to date and it was mentioned that all the changes to the specification are documented here: https://github.com/dotnet/runtime/blob/master/docs/design/specs/Ecma-335-Augments.md

But it seems that this file has not been updated for the changes made as part of the work for default interface methods:
https://github.com/dotnet/runtime/blob/master/docs/design/features/default-interface-methods.md

I thought I'd raise this issue just to see if the augments document should be updated with all the ECMA-335 changes that are missing. I believe there might have also been changes needed for supporting unmanaged calling conventions?

### PR DESCRIPTION
**Pull request by davidwrighton**
Update the ECMA augments file to reflect changes made in .NET 5.0 for covariant returns.

Fixes #37265
Fixes #43814

### CODE REVIEW

File: docs/design/specs/Ecma-335-Augments.md [hunk 1]
```diff
@@ -810,3 +820,86 @@ Bullet 2: It is valid to call a virtual method using `call` (rather than `callvi
...
+| ---  | --- | --- |
+| A | A::VirtualFunction() | A::VirtualFunction |
+| B | A::VirtualFunction() | B::VirtualFunction |
+| B | B::VirtualFunction() | B::VirtualFunction |
+| C | A::VirtualFunction() | C::VirtualFunction |
+| C | B::VirtualFunction() | C::VirtualFunction |
+| C | C::VirtualFunction() | C::VirtualFunction |
+"
+
+Add a new paragraph "When a rules above result in the implementation of virtual method not satisfying the requirement that the signature of the implementor of a virtual method is not *covariant-return-compatible-with* (§I.8.7.1) the program is not considered well formed. As an example, this can happen if a virtual method is overriden with a .override directive(A) paired with a `PreserveBaseOverridesAttribute` with a *covariant-return-compatible-with* signature. and then the initial virtual method is subsequently overridden with a method which is *covariant-return-compatible-with* the initial virtual method, but not with the method body associated with .override directive(A)."
```
Comment (janvorli) [inline docs/design/specs/Ecma-335-Augments.md:900]: A nit "When a rules" -> "When rules"

File: docs/design/specs/Ecma-335-Augments.md [hunk 2]
```diff
@@ -810,3 +820,86 @@ Bullet 2: It is valid to call a virtual method using `call` (rather than `callvi
...
+
+### II.22.27
+Edit rule 12 to specify that "The method isgnature defined by *MethodBody* shall match those defined by *MethodDeclaration* exactly if *MethodDeclaration* defines a method on an interface or be covariant-return-compatible-with* (§I.8.7.1) if *MethodDeclaration* represents a method on a class."
```
Comment (janvorli) [inline docs/design/specs/Ecma-335-Augments.md:903]: isgnature -> signature

Comment (janvorli) [reply docs/design/specs/Ecma-335-Augments.md:903]: Missing * in front of the covariant-return-compatible-with

File: docs/design/specs/Ecma-335-Augments.md [hunk 3]
```diff
@@ -810,3 +820,86 @@ Bullet 2: It is valid to call a virtual method using `call` (rather than `callvi
...
+
+class Derived : Base
+{
+  // Note that the VirtualMethod here overrides the VirtualMethod defined on Base even though the 
+  // return types are not the same. However, the return types are compatible in a covariant fashion.
+  public override string VirtualMethod() { return null;}
+}
+```
+
+Covariant return methods can only be described through MethodImpl records, and as are only be applicable to methods on reference types. Methods on interfaces and value types are not supported. In addition the covariant override is only applicable for overriding methods also defined on a reference type.
```
Comment (janvorli) [inline docs/design/specs/Ecma-335-Augments.md:845]: "and as are only ..." - was it meant to be "and as such are only ..."?

File: docs/design/specs/Ecma-335-Augments.md [hunk 4]
```diff
@@ -810,3 +820,84 @@ Bullet 2: It is valid to call a virtual method using `call` (rather than `callvi
...
+Add a new signature type relation.
+
+"A method signature type T is *covariant-return-compatible-with* with a method signature type U if and only if:
+1. The calling conventions of T and U shall match exactly, ignoring the distinction between static and instance methods (i.e., the this parameter, if any, is not treated specially).
+2. For each parameter type of P of T, and corresponding type Q of U, P is identical Q.
+3. For the return type P of T, and return type Q of U, Q is *compatible-with* P and is a *reference type* OR Q is identical to P.
+
+### II.10.3.2 The .override directive (page 147)
+
+Edit the first paragraph "The .override directive specifies that a virtual method shall be implemented (overridden), in this type, by a virtual method with a different name, but with a *covariant-return-compatible-with* signature (§I.8.7.1). This directive can be used to provide an implementation for a virtual method inherited from a base class, or a virtual method specified in an interface implemented by this type and all other. If not used to provide an implementation for a virtual method inherited from a base class the signature must be the identical."
```
Comment (trylek) [inline docs/design/specs/Ecma-335-Augments.md:864]: Nit - "the signature must be the identical" - as you know I'm not a native English speaker but should this rather be *the signature must be identical* or alternatively *the signature must be the same*?

File: docs/design/specs/Ecma-335-Augments.md [hunk 5]
```diff
@@ -810,3 +820,84 @@ Bullet 2: It is valid to call a virtual method using `call` (rather than `callvi
...
+
+### II.10.3.4 Impact of overrides on derived classes
+Add a third bullet
+"- If a virtual method is overridden via a .override directive and virtual method in parent class was overriden with a .override directive where the method body of that second .override directive is decorated with `System.Runtime.CompilerServices.PreserveBaseOverridesAttribute` then the virtual method override shall apply to the method body of the second .override directive as well.
```
Comment (trylek) [inline docs/design/specs/Ecma-335-Augments.md:868]: Nit - "an .override directive"?

File: docs/design/specs/Ecma-335-Augments.md (hunk 1 above)
Comment (trylek) [reply docs/design/specs/Ecma-335-Augments.md]: (or maybe *the rules*)

Comment (trylek) [reply docs/design/specs/Ecma-335-Augments.md]: (or *a rule* even though I'm not sure whether there's a real difference between the singular and plural in this case)

Comment (trylek) [reply docs/design/specs/Ecma-335-Augments.md]: Actually, I think there's also a confusing double negative in the sentence unless I'm reading it wrong - "something results in the implementation of virtual method not satisfying that the signature of something is not covariant compatible", at the very least it's quite a bit to chew on which is the traditional hell of all specs :-).

**Review by trylek — APPROVED** [review]
Looks great to me modulo a few nits, thanks David!

File: docs/design/specs/Ecma-335-Augments.md (hunk 4 above)
Comment (davidwrighton) [reply docs/design/specs/Ecma-335-Augments.md:864]: Sigh. My reading tends to gloss over that sort of issue, which is why a non-English speaker is actually quite helpful in this kind of review as you don't do that. Yes. the signature must be identical sounds right to me.

File: docs/design/specs/Ecma-335-Augments.md [hunk 6]
```diff
@@ -810,3 +820,84 @@ Bullet 2: It is valid to call a virtual method using `call` (rather than `callvi
...
+
+class Derived : Base
+{
+  // Note that the VirtualMethod here overrides the VirtualMethod defined on Base even though the 
+  // return types are not the same. However, the return types are compatible in a covariant fashion.
+  public override string VirtualMethod() { return null;}
+}
+```
+
+Covariant return methods can only be described through MethodImpl records, and are only be applicable to methods on non-interface reference types. In addition the covariant override is only applicable for overriding methods also defined on a non-interface reference type.
```
Comment (janvorli) [inline docs/design/specs/Ecma-335-Augments.md:845]: One last: "are only be applicable" -> "are only applicable"

Comment (trylek) [reply docs/design/specs/Ecma-335-Augments.md:845]: Nice, I missed that one :D.

**Comment by davidwrighton** [comment]
@trylek @janvorli I rewrote the error case to be part of the bullet point and added an example. I think that makes it a fair bit clearer.

**Review by janvorli — APPROVED** [review]
LGTM, thank you!

File: docs/design/specs/Ecma-335-Augments.md [hunk 7]
```diff
@@ -885,19 +886,25 @@ Add a third bullet
...
+| ---  | --- | --- | --- |
+| A | A::VirtualFunction() | A::VirtualFunction | |
+| B | A::VirtualFunction() | B::VirtualFunction | |
+| B | B::VirtualFunction() | B::VirtualFunction | |
+| C | A::VirtualFunction() | C::VirtualFunction | |
+| C | B::VirtualFunction() | C::VirtualFunction | |
+| C | C::VirtualFunction() | C::VirtualFunction | |
+| D | B::VirtualFunction() | ERROR | A program containing type D is not valid, as B::VirtualFunction would be implemented by D::VirtualFunction which is not *covariant-return-compatible-with* (§I.8.7.1) B::VirtualFunction |
 "
-
```
Comment (trylek) [inline docs/design/specs/Ecma-335-Augments.md]: Wow, I didn't expect such a profound effect on the wording, I really appreciate your diligence towards polishing the spec!

**Pull request merged**